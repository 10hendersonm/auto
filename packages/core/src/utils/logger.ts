import signale, { Signale } from 'signale';

export interface ILogger {
  /** The level at which to log messages */
  logLevel: LogLevel;
  /** The default logger. Always on */
  log: signale.Signale<signale.DefaultMethods>;
  /** The verbose log. Has more debug logs */
  verbose: signale.Signale<signale.DefaultMethods>;
  /** The very verbose log. Has all debug logs */
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

/** Create a dummy logger for testing. */
export function dummyLog(): ILogger {
  return {
    logLevel: undefined,
    log: new Signale({ disabled: true }),
    verbose: new Signale({ disabled: true }),
    veryVerbose: new Signale({ disabled: true })
  };
}

export type LogLevel = 'verbose' | 'veryVerbose' | undefined;

/** Create a logger the the given log level. */
export default function createLog(mode: LogLevel): ILogger {
  return {
    logLevel: mode,
    log: new Signale(),
    verbose: new Signale({
      disabled: mode !== 'verbose' && mode !== 'veryVerbose'
    }),
    veryVerbose: new Signale({
      disabled: mode !== 'veryVerbose'
    })
  };
}
