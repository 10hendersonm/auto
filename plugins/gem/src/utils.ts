import * as fs from "fs";
import { promisify } from "util";

export const mkdir = promisify(fs.mkdir);
export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);
