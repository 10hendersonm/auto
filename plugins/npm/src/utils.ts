import * as fs from "fs";
import { promisify } from "util";

export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);

/** Check if the project is a monorepo */
export const isMonorepo = () => fs.existsSync("lerna.json");

/** Parse the lerna.json file. */
export const getLernaJson = () => {
  try {
    return JSON.parse(fs.readFileSync("lerna.json", "utf8"));
  } catch (error) {
    return {};
  }
};
