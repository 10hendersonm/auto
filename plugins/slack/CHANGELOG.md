# v9.22.4 (Wed Apr 01 2020)

#### üêõ Bug Fix

- use release URL from response for an accurate URL [#1096](https://github.com/intuit/auto/pull/1096) ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v9.21.1 (Fri Mar 27 2020)

#### üêõ Bug Fix

- Respect Updated PR Titles [#1082](https://github.com/intuit/auto/pull/1082) ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v9.15.7 (Tue Mar 03 2020)

#### üêõ Bug Fix

- add license to all sub-packages ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v9.15.3 (Mon Mar 02 2020)

#### üêõ Bug Fix

- Bump @types/node-fetch from 2.5.4 to 2.5.5

Bumps [@types/node-fetch](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/node-fetch) from 2.5.4 to 2.5.5.

- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/node-fetch)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com> ([@dependabot-preview[bot]](https://github.com/dependabot-preview[bot]))

- Bump tslib from 1.11.0 to 1.11.1

Bumps [tslib](https://github.com/Microsoft/tslib) from 1.11.0 to 1.11.1.

- [Release notes](https://github.com/Microsoft/tslib/releases)
- [Commits](https://github.com/Microsoft/tslib/compare/1.11.0...1.11.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com> ([@dependabot-preview[bot]](https://github.com/dependabot-preview[bot]))

#### Authors: 1

- [@dependabot-preview[bot]](https://github.com/dependabot-preview[bot])

---

# v9.15.0 (Sun Mar 01 2020)

#### üêõ Bug Fix

- add fp and io-ts as deps ([@hipstersmoothie](https://github.com/hipstersmoothie))
- add plugin configuration validation ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v9.13.1 (Mon Feb 24 2020)

#### üêõ Bug Fix

- Bump tslib from 1.10.0 to 1.11.0

Bumps [tslib](https://github.com/Microsoft/tslib) from 1.10.0 to 1.11.0.

- [Release notes](https://github.com/Microsoft/tslib/releases)
- [Commits](https://github.com/Microsoft/tslib/compare/1.10.0...1.11.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com> ([@dependabot-preview[bot]](https://github.com/dependabot-preview[bot]))

#### Authors: 1

- [@dependabot-preview[bot]](https://github.com/dependabot-preview[bot])

---

# v9.6.0 (Mon Jan 27 2020)

#### üêõ Bug Fix

- add ability to store slack hook url in environment variable ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v9.5.0 (Mon Jan 27 2020)

#### üêõ Bug Fix

- fix lint ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Overhaul "auto init" experience + make it pluggable ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v8.8.0 (Thu Jan 02 2020)

#### üêõ Bug Fix

- test: added a test to assure prerelease branches don't trigger a slack ([@Aghassi](https://github.com/Aghassi))
- refactor: ensure the fetch call is never made ([@Aghassi](https://github.com/Aghassi))
- refactor: fixed incorrect logic for prerelease check ([@Aghassi](https://github.com/Aghassi))
- refactor: removed unecessary console logs ([@Aghassi](https://github.com/Aghassi))
- test: added tests for when publishPreRelease is set ([@Aghassi](https://github.com/Aghassi))
- docs: updated slack plugin docs on publishPreRelease usage ([@Aghassi](https://github.com/Aghassi))
- fix: prevent slack from default publishing on prerelease branches ([@Aghassi](https://github.com/Aghassi))
- refactor: removed ternary per @hipstersmoothie feedback

Co-Authored-By: Andrew Lisowski <lisowski54@gmail.com> ([@Aghassi](https://github.com/Aghassi))

#### Authors: 1

- David ([@Aghassi](https://github.com/Aghassi))

---

# v8.6.0 (Tue Dec 17 2019)

#### üêõ Bug Fix

- Bump version to: v8.5.0 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Update CHANGELOG.md \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v8.5.0 (Tue Dec 17 2019)

#### üêõ Bug Fix

- Bump version to: v8.4.0 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.3.0 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.2.0 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.1.3 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.1.2 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.1.1 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v8.0.0 (Wed Dec 11 2019)

#### üêõ Bug Fix

- Bump version to: v8.0.0-next.4 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- do not call afterRelease hooks during dry run ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.8 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.7 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- change skip-release releaseType to just skip ([@hipstersmoothie](https://github.com/hipstersmoothie))
- add none release type ([@hipstersmoothie](https://github.com/hipstersmoothie))
- get label refactor working ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.6 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.5 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- reset versions ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.3 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.2 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.1 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- fix tests ([@hipstersmoothie](https://github.com/hipstersmoothie))
- allow user to configure what branches are treated as prerelease branches ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v8.0.0-next.0 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v7.17.0-next.1 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Bump version to: v7.17.0-next.0 \[skip ci\] ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))
