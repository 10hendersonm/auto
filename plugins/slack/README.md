# Slack Plugin

Post your release notes to a slack channel

## Installation

This plugin is not included with the `auto` CLI installed via NPM. To install:

```sh
npm i --save-dev @intuit-auto/slack
# or
yarn add -D @intuit-auto/slack
```

## Usage

To use the plugin include it in your `.autorc`

```json
{
  "plugins": [
    ["slack", { "url": "https://url-to-your-slack-hook.com" }],
    // or
    ["slack", "https://url-to-your-slack-hook.com"]
  ]
}
```

This URL should be to you webhook. If you require a token to post to a slack hook, make sure you have a SLACK_TOKEN variable available on your environment. This token will be added to eh URL as a query string parameter.
