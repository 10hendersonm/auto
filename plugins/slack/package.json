{
  "name": "@auto-it/slack",
  "version": "7.3.3",
  "main": "dist/index.js",
  "description": "Slack plugin for auto",
  "author": {
    "name": "Andrew Lisowski",
    "email": "lisowski54@gmail.com"
  },
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/intuit/auto"
  },
  "files": [
    "dist"
  ],
  "keywords": [
    "automation",
    "semantic",
    "release",
    "github",
    "labels",
    "automated",
    "continuos integration",
    "changelog",
    "slack"
  ],
  "scripts": {
    "build": "tsc -p tsconfig.json",
    "build:watch": "npm run build -- -w",
    "lint": "tslint -p . --format stylish",
    "test": "jest --maxWorkers=2 --config ../../package.json"
  },
  "dependencies": {
    "@atomist/slack-messages": "~1.1.0",
    "@auto-it/core": "link:../../packages/core",
    "node-fetch": "2.5.0",
    "url-join": "^4.0.0"
  },
  "devDependencies": {
    "@types/node-fetch": "2.5.0",
    "@types/url-join": "^4.0.0"
  }
}
