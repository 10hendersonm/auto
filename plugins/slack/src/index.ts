import { githubToSlack } from '@atomist/slack-messages';
import { Auto, IPlugin } from '@auto-it/core';
import fetch from 'node-fetch';
import join from 'url-join';

const sanitizeMarkdown = (markdown: string) =>
  githubToSlack(markdown)
    .split('\n')
    .map(line => {
      // Strip out the ### prefix and replace it with *<word>* to make it bold
      if (line.startsWith('#')) {
        return `*${line.replace(/^[#]+/, '')}*`;
      }

      return line;
    })
    .join('\n');

interface ISlackPluginOptions {
  url: string;
  atTarget?: string;
}

export default class SlackPlugin implements IPlugin {
  name = 'Slack';

  readonly options: ISlackPluginOptions;

  constructor(options: ISlackPluginOptions | string) {
    if (typeof options === 'string') {
      this.options = { url: options, atTarget: 'channel' };
    } else {
      this.options = {
        url: options.url ? options.url : '',
        atTarget: options.atTarget ? options.atTarget : 'channel'
      };
    }
  }

  apply(auto: Auto) {
    auto.hooks.afterRelease.tapPromise(
      this.name,
      async ({ newVersion, commits, releaseNotes }) => {
        if (!newVersion) {
          return;
        }

        if ('dryRun' in auto.options && auto.options.dryRun) {
          return;
        }

        const head = commits[0];

        if (!head) {
          return;
        }

        const isSkipped = head.labels.find(label =>
          auto.release!.options.skipReleaseLabels.includes(label)
        );

        if (isSkipped) {
          return;
        }

        if (!this.options.url) {
          throw new Error('Slack url must be set to post a message to slack.');
        }

        await this.postToSlack(auto, newVersion, releaseNotes);
      }
    );
  }

  async postToSlack(auto: Auto, newVersion: string, releaseNotes: string) {
    if (!auto.git) {
      return;
    }

    auto.logger.verbose.info('Posting release notes to slack.');

    const project = await auto.git.getProject();
    const body = sanitizeMarkdown(releaseNotes);
    const token = process.env.SLACK_TOKEN;
    const releaseUrl = join(project.html_url, 'releases/tag', newVersion);
    const atTarget = this.options.atTarget;

    if (!token) {
      auto.logger.verbose.warn('Slack may need a token to send a message');
    }

    await fetch(`${this.options.url}${token ? `?token=${token}` : ''}`, {
      method: 'POST',
      body: JSON.stringify({
        text: [
          `@${atTarget}: New release *<${releaseUrl}|${newVersion}>*`,
          body
        ].join('\n'),
        // eslint-disable-next-line camelcase
        link_names: 1
      }),
      headers: { 'Content-Type': 'application/json' }
    });

    auto.logger.verbose.info('Posted release notes to slack.');
  }
}
