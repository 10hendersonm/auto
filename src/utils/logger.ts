import signale, { Signale } from 'signale';

export interface ILogger {
  log: signale.Signale<signale.DefaultMethods>;
  verbose: signale.Signale<signale.DefaultMethods>;
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

export function dummyLog(): ILogger {
  return {
    log: new Signale({ disabled: true }),
    verbose: new Signale({ disabled: true }),
    veryVerbose: new Signale({ disabled: true })
  };
}

export type LogLevel = 'verbose' | 'veryVerbose' | undefined;

export default function createLog(mode: LogLevel): ILogger {
  return {
    log: new Signale(),
    verbose: new Signale({
      disabled: mode !== 'verbose' && mode !== 'veryVerbose'
    }),
    veryVerbose: new Signale({
      disabled: mode !== 'veryVerbose'
    })
  };
}
