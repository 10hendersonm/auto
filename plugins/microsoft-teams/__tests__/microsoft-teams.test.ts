import { dummyLog } from "@auto-it/core/src/utils/logger";
import { sanitizeMarkdown } from "@auto-it/slack";
import createHttpsProxyAgent from "https-proxy-agent";

import MicrosoftTeamsPlugin from "../src";

const fetchSpy = jest.fn();
// @ts-ignore
jest.mock("node-fetch", () => (...args) => {
  fetchSpy(...args);
});

beforeEach(() => {
  fetchSpy.mockClear();
});

const mockAuto = {
  git: {},
  logger: dummyLog(),
} as any;

describe("postToMicrosoftTeams", () => {
  test("should call microsoft-teams api with minimal config", async () => {
    const plugin = new MicrosoftTeamsPlugin("https://microsoft-teams-url");

    await plugin.createPost(
      mockAuto,
      sanitizeMarkdown("# My Notes\n- PR [some link](google.com)"),
      "*<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*",
      undefined
    );

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe("https://microsoft-teams-url");
    expect(fetchSpy.mock.calls[0][1].agent).toBeUndefined();
    expect(fetchSpy.mock.calls[0][1].body).toMatchSnapshot();
  });

  test("should call microsoft-teams api through http proxy", async () => {
    const plugin = new MicrosoftTeamsPlugin("https://microsoft-teams-url");
    process.env.http_proxy = "http-proxy";

    await plugin.createPost(
      mockAuto,
      sanitizeMarkdown("# My Notes\n- PR [some link](google.com)"),
      "*<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*",
      createHttpsProxyAgent("mock-url")
    );

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe("https://microsoft-teams-url");
    expect(fetchSpy.mock.calls[0][1].agent).not.toBeUndefined();
    expect(fetchSpy.mock.calls[0][1].body).toMatchSnapshot();
  });
});
