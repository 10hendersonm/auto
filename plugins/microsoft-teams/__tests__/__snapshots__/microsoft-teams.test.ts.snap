// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postToMicrosoftTeams should call microsoft-teams api through http proxy 1`] = `"{\\"@context\\":\\"http://schema.org/extensions\\",\\"@type\\":\\"MessageCard\\",\\"text\\":\\"@channel: New release *<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\\\n* My Notes*\\\\n• PR <google.com|some link>\\"}"`;

exports[`postToMicrosoftTeams should call microsoft-teams api with minimal config 1`] = `"{\\"@context\\":\\"http://schema.org/extensions\\",\\"@type\\":\\"MessageCard\\",\\"text\\":\\"@channel: New release *<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\\\n* My Notes*\\\\n• PR <google.com|some link>\\"}"`;
